README
#==============================================================
#
#  Name: README
#
#  Rem:  
#        EPICS IOC Application 
#
#  Side: None
#
#  Auth: XX-Mar-2012,          (XX)
#  Rev:  dd-mmm-yyyy, Reviewer's Name (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        dd-mmm-yyyy, Reviewer's Name (USERNAME)
#          coment
#
#==============================================================
#

# End of file






ChangeLog
20160610 (T.S.):
 - sis8300Digi.c: print error message if register read/write ioctl fails
20150520 (T.S.):
 - sis8300Digi.h, sis8300.c: new option to let sis8300DigiValidateSel()
   run quietly.
 - sis8300Digi.h, sis8300.c: new call 'sis8300DigiGetFeatures() to inquire
   about features. So far only the 'number of channels' is supported.
 - sis8300Digi.c: added file-descriptor validity checks
 - c109.c:
     added '-c sel' option to let user define channel selector
     BUGFIX: default selector was set to all (10) channels which
             failed on 8-channel models. We now use the new 'sis8300DigiGetFeatures()'
             call to obtain the number of channels and clip the default
             selector accordingly.
20140320 (T.S.):
 - sis8300-R1-0-0 RELEASE
20140310 (T.S.):
 - sis8300/sis8300.c, sis8300/sis8300.h: added 'fd' argument
   to sis8300DigiValidateSel() since there are now models with only 8 ch.
   Having access to 'fd' the code can figure out the supported number
   of channels.
 - sis8300/sis8300.c: sis8300DigiSetCount() - BUGFIX: channel-disable
   mask was incorrectly computed if CH9 or CH10 were unused (both
   of these were always enabled and could corrupt readings of other
   channels in memory).
20131212 (T.S.):
 - sis8300/sis8300Digi.c, sis8300/sis8300Digi.h: added entry points
   si5326_status()
20131206 (T.S.):
 - sis8300/sis8300Digi.c sis8300/sis8300Digi.h: added entry points
   (sis8300Set9510Divider(), sis8300Get9510Clkhl()) to manipulate
   the 9510 divider ratio.
20131105 (T.S.):
 - sis8300/sis3800Digi.c:
    o fixed sis8300BuildChannelSel() (truncation to 32-bit issues)
    o do a few retries when waiting for Si5326 to lock
20131101 (T.S.):
 - sis8300Digi.c: 
    o when reading/writing ADC via SPI wait until not busy.
    o only set tap delay for supported of # channels
    o FIX: wait until tap dealy register *NOT* busy
 - added saio.c, saio.h (wrappers for linux' AIO syscalls)
 - made si5326_setup(int fd, Si5326Parms p) public
 - added sis8300DigiGetADC_ID(int fd)
 - added sis8300DigiGetFclkMax(int fd)
 - sis8300DigiSetup(): initially set MUX to internal clock and
   AD9510 to divide by 32; check whether the max. clock of the
   ADC is compatible with the requested clock frequency and
   reject if necessary. Program the final mux. and 9510 settings
   only if frequency check passes.



sis8300/c109.c
#include <inttypes.h>
#include <fcntl.h>
#include <stdio.h>
#include <sys/ioctl.h>
#include <getopt.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>

#include <sis8300Digi.h>

static void usage(const char *nm)
{
	fprintf(stderr,"Usage: %s [-d device] [-f freq] [-L loop_bandwidth] [-qh] [-c sel] [-S] [-b] [-B] [-N nblks] [-4] [-T W|N] [-C] <config>\n\n", nm);
	fprintf(stderr,"           -h         : print this message\n");
	fprintf(stderr,"           -q         : query Si5236 operating mode only\n");
	fprintf(stderr,"           -d device  : use 'device' (path to dev-node)\n");
	fprintf(stderr,"           -S         : set muxes to use si5326 clock. Note\n");
	fprintf(stderr,"                        that this (for legacy reasons) also\n");
	fprintf(stderr,"                        sets the frequency to 109MHz. Thus,\n");
	fprintf(stderr,"                        when you use -f either omit -S or\n");
	fprintf(stderr,"                        put -S before -f on the command line.\n");

	fprintf(stderr,"           -b         : do not bypass 9510 dividers (only if -S in wide-band mode)\n");
	fprintf(stderr,"           -B         : enforce bypass of 9510 dividers\n");
	fprintf(stderr,"           -e         : disable external trigger (enabled by default)\n");
	fprintf(stderr,"           -N nblks   : number of sample blocks (16samples) per channel\n");
	fprintf(stderr,"                        - defaults to 2.\n");
	fprintf(stderr,"           -4         : use channels 2,4,6,8 only\n");
	fprintf(stderr,"           -f freq    : program Si5326 for output frequency 'freq'\n");
	fprintf(stderr,"                        (implies -S)\n");
	fprintf(stderr,"           -T W|N     : only compute divider settings w/o accessing the device.\n");
	fprintf(stderr,"                        Requires '-f'. The user must specify the device mode\n");
	fprintf(stderr,"                        ('W'ide- or 'N'arrow-band).\n");
	fprintf(stderr,"           -L bw      : Set PLL loop bandwidth\n");
	fprintf(stderr,"           -C         : read config parameters <n3> <n2h> <n2l> <n1h> <nc> <bwsel>\n");
	fprintf(stderr,"           -I         : ignore fixed, hard-configured configurations\n");
	fprintf(stderr,"           -c sel     : provide selector which defines channel enablement and assignment\n");
	fprintf(stderr,"                        to memory layout (see sis8300Digi.h for more information)\n");
	fprintf(stderr,"           -v         : be verbose\n");
}

typedef struct {
	unsigned long    fout;
	Si5326ParmsRec   parms;
} Si5326ConfigRec, *Si5326Config;

static Si5326ConfigRec si5326Configs_none[] = {
	{
	fout: 0,
	parms: {}
	}
};

static Si5326ConfigRec si5326Configs_wb[] = {
	{ /* 109MHz wideband */
	fout: 109000000UL,
	parms:
	{
	  fin  : 250000000UL,
	  n3   : 10,
	  /* f3 = 25Mhz */
	  n2h  : 1,
	  n2l  : 109*2,
	  /* fo = 50*109 MHz */
	  n1h  : 5,
	  nc   : 10,
	  bwsel: 1,
	  wb   : 1,
	}
	},

	{ /* Use 500MHz wideband */
	fout: 500000000UL,
	parms:
	{
	  fin  : 250000000UL,
	  n3   : 2,
	  /* f3 = 125Mhz */
	  n2h  : 1,
	  n2l  : 44,
	  /* fo = 44*125 MHz */
	  n1h  : 11,
	  nc   : 1,
	  bwsel: 2,
	  wb  : 1,
	}
	},

	{ /* END marker */
	fout: 0,
	parms:
	{
	}
	}
};

static Si5326ConfigRec si5326Configs_nb[] = {
	{ /* 218MHz narrow-band */
	fout: 218000000UL,
	parms:
	{
	  fin  : 250000000UL,
	  n3   : 125,
	  /* f3 = 2Mhz */
	  n2h  : 4,
	  n2l  : 654,
	  /* fo = 2*109 MHz */
	  n1h  : 6,
	  nc   : 4,
	  bwsel: 5,
	  wb   : 0,
	}
	},

	{ /* END marker */
	fout: 0,
	parms:
	{
	}
	}
};

int
main(int argc, char **argv)
{
int      rval = 1;
int      fd = -1;
int      nblks = 2;
int      exttrig = 1;
int      enf_byp = 0;
int      query = 0;
int      do_config = 0;
int      ignore_fixed = 0;
int      i;
int      verbose   = 0;
unsigned div_clkhl = SIS8300_BYPASS_9510_DIVIDER;
int      opt;
const char *dev = getenv("RACC_DEV");
int      *i_p;
unsigned long *ul_p;
unsigned long freq = 0;
int      bw=0;
Si5326Parms        si5326_clk = 0;
Si5326Config       si5326_cfg = 0;
Sis8300ChannelSel  sel = 0xa987654321ULL;
Si5326Mode         mode = Si5326_Error;
Si5326ParmsRec     parms;
unsigned *pp[6];
uint64_t fout, rat;
unsigned long long *ull_p;
unsigned long long sel_i;
int      sel_i_set = 0;
long     f;

	while ( (opt = getopt(argc, argv, "hqSbBed:N:4f:CT:IvL:c:")) > 0 ) {
		i_p   = 0;
		ul_p  = 0;
		ull_p = 0;
		switch ( opt ) {
			default:
			case 'h': usage(argv[0]); return 0;

			case 'S': freq = 109000000UL; break;
			case 'b': enf_byp    = -1; break;
			case 'B': enf_byp    =  1; break;
	
			case 'd': dev        = optarg; break;

			case 'e': exttrig    = 0; break;

			case 'N': i_p        = &nblks; break;
	
			case '4': sel = 0x8642ULL; sel_i_set = 0; break;

			case 'f': ul_p = &freq; break;
			case 'C': do_config = 1; break;
			case 'q': query = 1; break;
			case 'v': verbose = 1; break;
			
			case 'T': switch ( toupper( (int)(unsigned char)optarg[0] ) ) {
			              default: fprintf(stderr,"Option -T needs 'W' or 'N' argument\n"); return 1;
			              case 'W': mode = Si5326_WidebandMode;   break;
			              case 'N': mode = Si5326_NarrowbandMode; break;
					  }
			          break;

			case 'L': i_p = &bw; break;

			case 'I': ignore_fixed = 1; break;

			case 'c': ull_p = &sel_i; sel_i_set = 1; break;
		}

		if ( i_p ) {
			if ( 1 != sscanf(optarg,"%i",i_p) ) {
				fprintf(stderr,"Option '%c' needs integer argument (scanning failed)!\n", opt);
				return 1;
			}
		}
		if ( ul_p ) {
			if ( 1 != sscanf(optarg,"%li",ul_p) ) {
				fprintf(stderr,"Option '%c' needs long integer argument (scanning failed)!\n", opt);
				return 1;
			}
		}
		if ( ull_p ) {
			if ( 1 != sscanf(optarg,"%lli",ull_p) ) {
				fprintf(stderr,"Option '%c' needs long long integer argument (scanning failed)!\n", opt);
				return 1;
			}
		}
	}

	parms.bw = bw;

	if ( do_config ) {
		unsigned bwsel;
		if ( Si5326_Error != mode ) {
			fprintf(stderr,"Cannot use both: -C and -T\n");
			return 1;
		}
		if ( freq > 0 ) {
			fprintf(stderr,"Cannot use both: -C and -f\n");
			return 1;
		}
		if ( optind + 5 > argc ) {
			fprintf(stderr,"Option -C needs 6 configuration parameters\n");
			return 1;
		}
		pp[0] = &parms.n3;
		pp[1] = &parms.n2h;
		pp[2] = &parms.n2l;
		pp[3] = &parms.n1h;
		pp[4] = &parms.nc;
		pp[5] = &bwsel;
		for ( i=0; i<6; i++ ) {
			if ( 1 != sscanf(argv[optind+i], "%u", pp[i]) ) {
				fprintf(stderr,"Option -C: unable to scan parameter %i\n", i+1);
				return 1;
			}
		}
		parms.bwsel = (int)bwsel;
		parms.fin   = 250000000UL;
		parms.wb    = 0;
	}

	if ( !freq )
		div_clkhl = 0;

	switch ( enf_byp ) {
		default:
		break;
		case -1: div_clkhl = 0; break;
		case +1: div_clkhl = SIS8300_BYPASS_9510_DIVIDER; break;
	}

	if ( Si5326_Error == mode ) {

		if ( ! dev ) {
			fprintf(stderr,"No device - use '-d <device>' or set RACC_DEV env_var\n");
			return 1;
		}

		if ( (fd = open(dev, O_RDWR)) < 0 ) {
			perror("opening device");
			return 1;
		}

	} else {
		if ( 0 == freq ) {
			fprintf(stderr, "if you use -T you must also use -f\n");
			return 1;
		}
		query = 0;
	}

	if ( freq > 0 || do_config || query ) {
		if ( Si5326_Error == mode ) {
			mode = sis8300ClkDetect( fd );
		}
		switch ( mode ) {
			default:
				fprintf(stderr,"Sis8300ClkDetect - unknown result %i\n", mode);
				return 1;

			case Si5326_NoReference:
				fprintf(stdout,"Si5326 - no reference detected\n");
				return 1;

			case Si5326_NarrowbandMode:
				if ( fd >= 0 ) {
					fprintf(stdout,"Si5326 - operating in narrow-band mode\n");
				}
				si5326_cfg = si5326Configs_nb;
				break;
			case Si5326_WidebandMode:
				if ( fd >= 0 ) {
					fprintf(stdout,"Si5326 - operating in wide-band mode\n");
				}
				si5326_cfg = si5326Configs_wb;
				parms.wb   = 1;
				break;
		}

		if ( query > 0 ) {
			/* query operating mode only */
			return 0;
		}

		if ( ignore_fixed )
			si5326_cfg = si5326Configs_none;

		if ( freq > 0 ) {
			for ( ; si5326_cfg->fout > 0; si5326_cfg++ ) {
				if ( freq == si5326_cfg->fout ) {
					div_clkhl = SIS8300_BYPASS_9510_DIVIDER;
					break;
				} else if ( 2*freq == si5326_cfg->fout ) {
					div_clkhl = 0;
					break;
				}
			}

			if ( 0 == si5326_cfg->fout ) {
				parms.fin = 250000000UL;
				parms.wb  = ( Si5326_WidebandMode == mode );
				if ( si53xx_calcParms( freq, &parms, verbose ) ) {
					fprintf(stderr, "Sorry, no configuration for output frequency %luHz found\n", freq);
					return 1;
				}
				si5326_clk = &parms;
			} else {
				si5326_clk = &si5326_cfg->parms;
			}
		}

		if ( do_config ) {
			si5326_clk = &parms;
		}
	}

	if ( fd >= 0 ) {

		if ( sis8300DigiSetup( fd, si5326_clk, div_clkhl, exttrig ) ) {
			goto bail;
		}

		if ( sel_i_set ) {
			sel = sel_i;
		} else {
			if ( (f = sis8300DigiGetFeatures( fd )) < 0 )
				goto bail;
			/* clip default selector to supported channels */
			sel &= (1ULL << (SIS8300_FEAT_N_CHANNELS(f) * 4)) - 1ULL;
		}

		if ( sis8300DigiSetCount(fd, sel, nblks * 16) ) {
			fprintf(stderr,"Setting sample count failed\n");
		}

	}
	if ( verbose || fd < 0 ) {
		if ( si5326_clk ) {
			fout = (uint64_t)si5326_clk->fin * (uint64_t)si5326_clk->n2h * (uint64_t)si5326_clk->n2l;
			fout /= (uint64_t)si5326_clk->n3 * (uint64_t)si5326_clk->n1h * (uint64_t)si5326_clk->nc;
			printf("PLL Input  Frequency:  %9luHz\n\n", si5326_clk->fin); 
			printf("                fin  %-4u*%4u\n", si5326_clk->n2h, si5326_clk->n2l);
			printf("Divider: fout = ---  ---------\n");
			printf("               %4u  %-4u*%4u\n\n", si5326_clk->n3, si5326_clk->n1h, si5326_clk->nc);
			printf("PLL Bandwidth:         %9uHz\n", si5326_clk->bw);
			printf("PLL Output Frequency:  %9"PRIu64"Hz\n", fout);
		} else {
			fout = 250000000ULL;
			printf("PLL Bypassed; Output Frequency %9"PRIu64"Hz\n", fout);
		}
        rat = div_clkhl > 0xff ? 1 : (div_clkhl & 0xf) + ((div_clkhl>>4) & 0xf) + 2;
		printf("AD9510 divider ratio:  %9"PRIu64"\n", rat);
		printf("Digitizer clock:       %9"PRIu64"Hz\n", fout/rat);
	}

	rval = 0;
bail:
	close( fd );
	return rval;
}


sis8300/saio.c
#include "saio.h"
#include <sys/syscall.h>
#include <stdint.h>
#include <string.h>

#define DEBUG

#ifdef DEBUG
#include <stdio.h>
#endif

/* Basic primitives */
int
saio_ctx_create(unsigned n_events, aio_context_t *ctxp)
{
	*ctxp = 0;
	return syscall(SYS_io_setup, n_events, ctxp);
}

int
saio_ctx_destroy(aio_context_t ctx)
{
	return syscall(SYS_io_destroy, ctx);
}

int
saio_submit(aio_context_t ctx, long n_iocbs, struct iocb **iocbparr)
{
	return syscall(SYS_io_submit, ctx, n_iocbs, iocbparr);
}

int
saio_cancel(aio_context_t ctx, struct iocb *iocb, struct io_event *result)
{
	return syscall(SYS_io_cancel, ctx, iocb, result);
}

int
saio_getevents(aio_context_t ctx, long min_nr, long nr, struct io_event *events, struct timespec *timeout)
{
	return syscall(SYS_io_getevents, ctx, min_nr, nr, events, timeout);
}
      

/* Helpers: */

/* Submit a single iocb for reading data */
void
saio_setup_for_read(struct iocb *iocb, int fd, void *buf, size_t count, off_t off)
{
	memset(iocb, 0, sizeof(*iocb));
	iocb->aio_lio_opcode = IOCB_CMD_PREAD;
	iocb->aio_fildes     = fd;
	iocb->aio_buf        = (uintptr_t)buf;
	iocb->aio_nbytes     = count;
	iocb->aio_offset     = off;
}

/* Use a iocb on the stack; this is essentially a synchronous read
 * but with a timeout
 */
ssize_t
saio_pread(aio_context_t ctx, int fd, void *buf, size_t count, off_t off, struct timespec *timeout)
{
struct iocb     b;
struct iocb    *ba[] = { &b };
int             st, rval;
struct io_event e;
	saio_setup_for_read(&b, fd, buf, count, off);
	if ( (st = saio_submit(ctx, sizeof(ba)/sizeof(ba[0]), ba)) < 0 ) {
#ifdef DEBUG
		perror("saio_submit ERROR");
#endif
		return st;
	}
	st = saio_getevents(ctx, 1, 1, &e, timeout);
#ifdef DEBUG
	if ( st <= 0 ) {
		if ( st < 0 )
			perror("saio_getevents ERROR");
		else
			fprintf(stderr,"saio_getevents TIMEOUT\n");
	}
#endif
	rval = e.res;
	if ( st <= 0 ) {
#ifdef DEBUG
		int st1 = 
#endif
		saio_cancel(ctx, &b, &e);
#ifdef DEBUG
		if ( st1 )
			perror("saio_cancel ERROR");
#endif
	}
	return st > 0 ? rval : st;
}

ssize_t
saio_pread_test(int fd, void *buf, size_t count, off_t off, struct timespec *timeout)
{
aio_context_t c = 0;
ssize_t       st;
#ifdef DEBUG
int           st1;
#endif

	if ( saio_ctx_create(1, &c) )  {
#ifdef DEBUG
		perror("saio_ctx_create ERROR");
#endif
		return -1;
	}

	st = saio_pread(c, fd, buf, count, off, timeout);	

#ifdef DEBUG
	st1 =
#endif
	saio_ctx_destroy( c );
#ifdef DEBUG
	if ( st1 )
		perror("saio_ctx_destroy ERROR");
#endif

	return st;
}


sis8300/saio.h
#ifndef SAIO_INCLUDE_H
#define SAIO_INCLUDE_H

/* Wrapper for linux' io_submit/io_setup/io_cancel etc. */
#include <unistd.h>
#include <linux/aio_abi.h>
#include <time.h>


/* Basic primitives */
int
saio_ctx_create(unsigned n_events, aio_context_t *ctxp);

int
saio_ctx_destroy(aio_context_t ctx);

int
saio_submit(aio_context_t ctx, long n_iocbs, struct iocb **iocbs);

int
saio_cancel(aio_context_t ctx, struct iocb *iocb, struct io_event *result);

int
saio_getevents(aio_context_t ctx, long min_nr, long nr, struct io_event *events, struct timespec *timeout);
      

/* Helpers: */

/* Submit a single iocb for reading data */
void
saio_setup_for_read(struct iocb *iocb, int fd, void *buf, size_t count, off_t off);

/* Use a iocb on the stack; this is essentially a synchronous read
 * but with a timeout
 */
ssize_t
saio_pread(aio_context_t ctx, int fd, void *buf, size_t count, off_t off, struct timespec *timeout);


#endif


sis8300/ratapp.h
#ifndef RATAPP_H
#define RATAPP_H

#include <inttypes.h>

/* Approximation of rational numbers with others that have a smaller
 * denominator.
 */
#ifndef TEST_SMALL
typedef uint64_t RatNum;

#define RATNUM_MAX ((uint64_t)(-1LL))
#define RATu PRIu64
#define RATx PRIx64
#else
typedef uint32_t RatNum;

#define RATNUM_MAX ((uint32_t)(-1LL))
#define RATu PRIu32
#define RATx PRIx32
#endif

typedef struct Rational_ {
	RatNum n,d;
} Rational;

#define CONV_N1 1
#define CONV_N2 0

typedef struct Convergent_ {
	RatNum      a;
	Rational    conv;
} Convergent;

/* NOTES:
 *  - routines dont' handle overflow gracefully
 *  - must not supply denominator or zero.
 *  - r_max specifies max. acceptable numerator and denominator. 
 *    Values of zero mean 'infinite' - but not both may be set
 *    to zero or RATNUM_MAX.
 */

/* find (up to) the last 'n' convergents of n/d with denominator smaller or equal
 * to r_max.d and numerator smaller or equal to r_max.n.
 * 
 * The user must pass an array of 'n' Convergent structs. If the continued
 * fraction terminates then the 'a' member of the last convergent is zero.
 * RETURNS: The index of the last convergent. The last convergent is c[k].
 * NOTES:   Storage wraps around, i.e., if k >= n then the index of
 *          c[j] with j associated with k, k-1, k-(n-1) is computed as j = k MOD n;
 * 
 *          I.e., the last convergent is always at c[k % n];
 */
int
ratapp_find_convergents(Convergent *c, int n, Rational *r_in, Rational *r_max);

/* like ratapp_find_convergents() but store only the last two convergents
 * RETURNS: index/order of last convergent.
 * NOTES:   c[0] = c(n-1)
 *          c[1] = c(n-1)
 *
 *          c[1].a is set to zero (not stored).
 */
int
ratapp_find_last_convergents(Convergent c[2], Rational *r_in, Rational *r_max);

/* Return the number of convergents for r_in which have r_in->d <= r_max->d
 * and r_in->n <= r_max->n.
 */
int
ratapp_estimate_terms(Rational *r_in, Rational *r_max);

/* Find best rational approximation with denominator smaller or equal to d_max */
void
ratapp_find_rational(Rational *r, Rational *r_in, Rational *r_max);

/* Compute 'best' approximation considering 'l-th' intermediate fraction.
 * For convenience: if l is 0 or >= c1->a then l is set to c1-a - 1 (since
 * l == c1->a would yield the next convergent.
 *
 * RETURNS: l > 0 if intermediate fraction was used, 0 if convergent c1 was
 *          better. Best approximation stored in *r.
 *
 * NOTE:    This may be used to iterate through convergents to find a best 
 *          approximation where other criteria need to be met:
 *
 *          RatNum l = 0;
 *          k        = ratapp_find_convergents(c, n, &r_in, &r_max);
 *
 *          // c[k] is last convergent which meets r_max. Hence
 *          // c[k-1]+a[k]*c[k] == c[k+1] is not acceptable.
 *
 *          while ( --k >= 0 ) {
 *              // iterate through all intermediates
 *              l = c[k+1].a;
 *              do {
 *                  l--;
 *          		l = ratapp_intermediate( &r, l, &c[k+1], &c[k], &r_in );
 *                  if ( acceptable( &r ) ) {
 *          			// terminate loops
 *                      goto done;
 *                  }
 *              } while ( l > 0 );
 *          }
 */
RatNum
ratapp_intermediate(Rational *r, RatNum l, Convergent *c1, Convergent *c2, Rational *r_in);

#endif


sis8300/sis8300Digi.h
#ifndef SIS8300DIGI_H
#define SIS8300DIGI_H

#include <stdint.h>

#define SIS8300_KIND_OFF  (-1)
#define SIS8300_KIND_BEAM 0
#define SIS8300_KIND_CRED 1
#define SIS8300_KIND_CGRN 2

typedef uint64_t Sis8300ChannelSel;

typedef struct Si5326Parms_ {
	unsigned long fin;
	unsigned n3, n2h, n2l, n1h, nc;
	unsigned bw;
	int      bwsel;
	int      wb;
} Si5326ParmsRec, *Si5326Parms;

typedef enum Si5326Mode_ {
	/* Error communicating with device; errno may be set */
	Si5326_Error          = -1,
	Si5326_NoReference    = 0,
	Si5326_NarrowbandMode = 1,
	Si5326_WidebandMode   = 2
} Si5326Mode;

/* Try to compute divider configuration for output frequency 'fout'.
 * Note that 'fout' is the si53xx output frequency may be further
 * divided by the 9510.
 *
 *
 * NOTE:    The user must set the PLL input frequency (250000000),
 *          the desired PLL bandwidth as well as the mode the device is
 *          using before calling this routine.
 *          The 'fout' parameter is in Hz.
 *
 * RETURNS: 0 on success, nonzero on error (*p does not contain
 *          valid settings in this case).
 *
 *          On success the 'n3, n2h, n2l, n1h, nc, bwsel' and
 *          realizable 'bw' are filled into *p by the routine.
 */
int
si53xx_calcParms(uint64_t fout, Si5326Parms p, int verbose);

/*
 * Program the si5326 with the given parameters
 *
 * RETURNS: -1 on error; output frequency (which may
 *          differ from the requested frequency due to
 *          rational approximation).
 */
int64_t
si5326_setup(int fd, Si5326Parms p);

/*
 * Obtain basic status of the si5326
 */
#define SIS8300_SI5326_NO_REF	(1<<0) /* no reference present */
#define SIS8300_SI5326_NO_CLK1  (1<<1) /* no signal at CLK1    */
#define SIS8300_SI5326_NO_CLK2  (1<<2) /* no signal at CLK2    */
#define SIS8300_SI5326_NO_LOCK  (1<<8) /* PLL not locked       */
int
si5326_status(int fd);

/* Probe the 5326 chip to find out if it has a usable narrow-band
 * reference, if it is strapped for wide-band mode or if there is
 * no valid reference (original Sis8300 module had Si5326 strapped
 * for ext. reference but none loaded).
 *
 * NOTE: This routine results in a SOFT-RESET of the Si5326 chip!
 */
Si5326Mode
sis8300ClkDetect(int fd);

/* Change the 9510 divider - clkhl is *not* the divider ratio
 * but the pattern of hi/lo times (consult the ad9510 datasheet)
 */
void
sis8300DigiSet9510Divider(int fd, unsigned clkhl);

/* Bypass the 9510 divider (div. ratio 0)             */
#define SIS8300_BYPASS_9510_DIVIDER 0xffffffff

/* Hold 9510 output steady low (div. ration infinite) */
#define SIS8300_SILENT_9510_DIVIDER 0xfffffffe

/*
 * Convert ratio to clkhl pattern.
 * Valid divider ratios are even numbers up to and including 32
 * as well as 1 (in which case the routine returns
 * SIS8300_BYPASS_9510_DIVIDER).
 * Odd ratios (different from 1) are rounded down; the result
 * for ratios bigger than 32 are undefined.
 */
unsigned
sis8300DigiGet9510Clkhl(unsigned ratio);


/* clkhl value to bypass the 9510 divider; the divider
 * is automatically engaged (div by 2) if no si5326_parms are
 * given.
 */
int
sis8300DigiSetup(int fd, Si5326Parms si5326_parms, unsigned clkhl, int exttrig_en);

/* channel_selector defines the order (and number) of channels in memory.
 * E.g., to have channels 4, 1, 8, 9 in this order in memory set 
 * 'channel_selector' = (9 << 12) | (8 << 8) | (1 << 4) | (4 << 0)
 * NOTE: channels are numbered 1..10; a 0 marks the end of the selected
 *       channels;
 *
 * 'nsmpl' defines samples per channel which MUST be a multiple of 16!
 */

/* Build a basic channel selector starting at channel 'start' up
 * to and including 'end', i.e., 
 *   (start<<0) | ((start+1)<<4) | ... | ((end)<<x)
 *
 * RETURNS selector or empty selector on error (start and end must be
 * 1..10, start <= end).
 */
Sis8300ChannelSel
sis8300BuildChannelSel(unsigned start, unsigned end);

/* Check the channel selector for validity.
 * For sake of backwards compatibility: if the
 * macro SIS8300_VALIDATE_SEL_QUIET is ORed into
 * the selector then no error messages are printed
 * if the selector is invalid. The caller may just
 * use the return value (-1).
 */ 

#define SIS8300_VALIDATE_SEL_QUIET (0x8000000000000000ULL)

int
sis8300DigiValidateSel(int fd, Sis8300ChannelSel sel);

int
sis8300DigiSetCount(int fd, Sis8300ChannelSel channel_selector, unsigned nsmpl);

int
sis8300DigiArm(int fd, int kind);

void
sis8300DigiSetSim(int fd, int32_t a, int32_t b, int32_t c, int32_t d, int quiet);

/* Optionally write qspi message to remote device and
 * optionally read response.
 * If 'data_out' non-negative then send message (data_out).
 * If data_in is not NULL then read response 
 * and store in *data_in.
 *
 * RETURNS: 0 on success
 *          nonzero (-1) if transfer timed out
 */
int 
sis8300DigiQspiWriteRead(const void *device, int data_out, uint16_t *data_in);

int
sis8300DigiReadReg(int fd, unsigned reg, uint32_t *val_p);

int
sis8300DigiWriteReg(int fd, unsigned reg, uint32_t val);

/* Retrieve chip ID of the first ADC chip
 *
 * RETURNS: chip ID or a negative number on error.
 */
int
sis8300DigiGetADC_ID(int fd);

/* Determine the max. clock frequency supported by the
 * first ADC chip on board.
 * 
 * RETURNS: max. clock frequency or zero on error,
 */
unsigned long
sis8300DigiGetFclkMax(int fd);

/* Set tap delay for fclk (Hz) -- it SUCKS that we have to to this */
void
sis8300DigiSetTapDelay(int fd, unsigned long fclk);

/* Obtain information about some of the features
 * of this digitizer. So far, only the 'number of channels'
 * is implemented.
 *
 * NOTE: before extracting any feature with the respective
 *       macros you MUST check the return value which
 *       may be -1 to indicate errors.
 */
long
sis8300DigiGetFeatures(int fd);

#define SIS8300_FEAT_N_CHANNELS(f) ((f)&0xf)

#endif
